File: ConnectFour.java
import java.util.Scanner;

public class ConnectFour {
    private static final int ROWS = 6;
    private static final int COLS = 7;
    private static final char EMPTY = '.';
    private static final char PLAYER_ONE = 'X';
    private static final char PLAYER_TWO = 'O';

    private char[][] board = new char[ROWS][COLS];

    public ConnectFour() {
        for (int i = 0; i < ROWS; i++) {
            for (int j = 0; j < COLS; j++) {
                board[i][j] = EMPTY;
            }
        }
    }

    public void printBoard() {
        for (int i = 0; i < ROWS; i++) {
            for (int j = 0; j < COLS; j++) {
                System.out.print(board[i][j] + " ");
            }
            System.out.println();
        }
        System.out.println("0 1 2 3 4 5 6");
    }

    public boolean makeMove(int col, char player) {
        for (int i = ROWS - 1; i >= 0; i--) {
            if (board[i][col] == EMPTY) {
                board[i][col] = player;
                return true;
            }
        }
        return false;
    }

    public boolean checkWin(char player) {
        // Check horizontal, vertical, and diagonal directions
        for (int row = 0; row < ROWS; row++) {
            for (int col = 0; col < COLS; col++) {
                if (checkDirection(row, col, 1, 0, player) || // Horizontal
                    checkDirection(row, col, 0, 1, player) || // Vertical
                    checkDirection(row, col, 1, 1, player) || // Diagonal \
                    checkDirection(row, col, 1, -1, player)) { // Diagonal /
                    return true;
                }
            }
        }
        return false;
    }

    private boolean checkDirection(int row, int col, int rowDir, int colDir, char player) {
        int count = 0;
        for (int i = 0; i < 4; i++) {
            int r = row + i * rowDir;
            int c = col + i * colDir;
            if (r >= 0 && r < ROWS && c >= 0 && c < COLS && board[r][c] == player) {
                count++;
            } else {
                break;
            }
        }
        return count == 4;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        ConnectFour game = new ConnectFour();
        char currentPlayer = PLAYER_ONE;
        boolean gameWon = false;

        while (true) {
            game.printBoard();
            System.out.println("Player " + currentPlayer + ", enter a column (0-6): ");
            int col = scanner.nextInt();

            if (col < 0 || col >= COLS || !game.makeMove(col, currentPlayer)) {
                System.out.println("Invalid move, try again.");
                continue;
            }

            if (game.checkWin(currentPlayer)) {
                game.printBoard();
                System.out.println("Player " + currentPlayer + " wins!");
                gameWon = true;
                break;
            }

            currentPlayer = (currentPlayer == PLAYER_ONE) ? PLAYER_TWO : PLAYER_ONE;
        }

        if (!gameWon) {
            System.out.println("It's a draw!");
        }

        scanner.close();
    }
}

File: ConnectFour.js
const ROWS = 6;
const COLS = 7;
const EMPTY = '.';
const PLAYER_ONE = 'X';
const PLAYER_TWO = 'O';

let board = Array.from({ length: ROWS }, () => Array(COLS).fill(EMPTY));

function printBoard() {
    console.clear();
    board.forEach(row => console.log(row.join(' ')));
    console.log('0 1 2 3 4 5 6');
}

function makeMove(col, player) {
    for (let i = ROWS - 1; i >= 0; i--) {
        if (board[i][col] === EMPTY) {
            board[i][col] = player;
            return true;
        }
    }
    return false;
}

function checkWin(player) {
    const directions = [
        { row: 0, col: 1 },  // Horizontal
        { row: 1, col: 0 },  // Vertical
        { row: 1, col: 1 },  // Diagonal \
        { row: 1, col: -1 }  // Diagonal /
    ];

    for (let row = 0; row < ROWS; row++) {
        for (let col = 0; col < COLS; col++) {
            if (board[row][col] === player) {
                for (let dir of directions) {
                    if (checkDirection(row, col, dir, player)) {
                        return true;
                    }
                }
            }
        }
    }
    return false;
}

function checkDirection(row, col, dir, player) {
    for (let i = 0; i < 4; i++) {
        const r = row + i * dir.row;
        const c = col + i * dir.col;
        if (r < 0 || r >= ROWS || c < 0 || c >= COLS || board[r][c] !== player) {
            return false;
        }
    }
    return true;
}

function playGame() {
    let currentPlayer = PLAYER_ONE;
    let gameWon = false;

    while (true) {
        printBoard();
        const col = prompt(`Player ${currentPlayer}, enter a column (0-6): `);
        if (col < 0 || col >= COLS || !makeMove(col, currentPlayer)) {
            console.log("Invalid move, try again.");
            continue;
        }

        if (checkWin(currentPlayer)) {
            printBoard();
            console.log(`Player ${currentPlayer} wins!`);
            gameWon = true;
            break;
        }

        currentPlayer = currentPlayer === PLAYER_ONE ? PLAYER_TWO : PLAYER_ONE;
    }

    if (!gameWon) {
        console.log("It's a draw!");
    }
}

playGame();
